### ISSUE TITLE
[NEXT-878] False positive image size warning for non-integer width/height

### ISSUE THREAD

**Issue by kamilmielnik** [vercel/next.js#47278]
### Verify canary release

* [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```
Operating System:
      Platform: linux
      Arch: x64
      Version: #36~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Fri Feb 17 15:17:25 UTC 2
    Binaries:
      Node: 16.19.1
      npm: 8.19.3
      Yarn: N/A
      pnpm: N/A
    Relevant packages:
      next: 13.2.5-canary.8
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

Image optimization (next/image, next/legacy/image)

### Link to the code that reproduces this issue

[https://codesandbox.io/p/sandbox/next-878-iyl8kd?file=%2Fpages%2Findex.js](https://codesandbox.io/p/sandbox/next-878-iyl8kd?file=%2Fpages%2Findex.js)

### To Reproduce

Providing non-integer `width` or `height` prop to `Image` component will always result in this warning:

```
Image with src "/_next/static/media/test.53caadc0.svg" has either width or height modified, but not the other. If you use CSS to change the size of your image, also include the styles 'width: "auto"' or 'height: "auto"' to maintain the aspect ratio.
```

### Describe the Bug

![](https://uploads.linear.app/099ed59a-06c6-472f-9977-4798721bda90/cbaa28d7-b3f2-4244-b6a5-73d8c8545580/3eb417af-12a5-46e5-92d2-9413aca6841d)

### Expected Behavior

Non-integer width/height props for Image component should not produce a warning

### Which browser are you using? (if relevant)

*No response*

### How are you deploying your application? (if relevant)

*No response*

<sub>From [SyncLinear.com](https://synclinear.com) | [NEXT-878](https://linear.app/vercel/issue/NEXT-878/false-positive-image-size-warning-for-non-integer-widthheight)</sub>

**Comment by balazsorban44** [comment]
Thanks for reporting. This comes from the following line:

https://github.com/vercel/next.js/blob/canary/packages/next/src/client/image.tsx/#L358

Where `img.width` rounds up to `11` for your reproduction.

**Comment by brunocroh** [comment]
@balazsorban44 I would like to help w/ this one, if it's not assigned to anyone already.


This will be my first PR, in next.js I thought of two solutions to this problem.

-  1st is to use [getItem](https://github.com/vercel/next.js/blob/3a49a29a7068ce7fb516f349d89416b99e4f8f87/packages/next/src/client/image.tsx#L257) to round, something like the example bellow

```ts
const heightModified =
  img.height.toString() !== getInt(img.getAttribute('height'))?.toString()
const widthModified =
  img.width.toString() !== getInt(img.getAttribute('width'))?.toString()
if (
  (heightModified && !widthModified) ||
  (!heightModified && widthModified)
) {
  warnOnce(
    `Image with src "${src}" has either width or height modified, but not the other. If you use CSS to change the size of your image, also include the styles 'width: "auto"' or 'height: "auto"' to maintain the aspect ratio.`
  )
}
```

- A 2nd option is to create a new warning when the value is not an integer.

What is the best approach to this issue? If you think something different, let me know!

**Comment by kamilmielnik** [comment]
My original expectation was that non-integer `width` & `height` props should not produce a warning.

Is there a reason why `width` & `height` need to be integers?

**Comment by kamilmielnik** [comment]
> But reading the [Image Component](https://github.com/vercel/next.js/blob/3a49a29a7068ce7fb516f349d89416b99e4f8f87/packages/next/src/client/image.tsx#L519), width & height need to be integer, to calculate the another when only one is passed.

Still, I don't see a reason for any of them to be integer.

In fact, forcing these values to be integers will result in a different aspect ratio (e.g. rounding `2.5px * 2px` to `2px * 2px` ), causing quality loss on paint.

This seems to be the original PR which introduced integer casting: https://github.com/vercel/next.js/commit/61fab923ebb2f77b88cfe13987771ad6e98eae08. The description or related issues don't indicate any need for integers.



For the record, non-integer values are useful e.g. for vector images (SVG).

**Comment by darincon** [comment]
I'm also experiencing this issue. This is causing problems too when the source image has odd numbers for dimensions. For example if the original image's dimensions are 1115 × 1355 and I want to display it with a width of 271px the height would be 239.331839px. I agree the solution shouldn't cast dimensions to integers. If browsers are capable and always work in decimals when resizing elements with height: auto, I don't see why NextJS has to do it

**Comment by gynekolog** [comment]
Yes, the issue is still relevant. And it's probably duplicity of https://github.com/vercel/next.js/issues/56025

**Comment by marcoshernanz** [comment]
Thanks for the report. This is the same false-positive image dimension warning tracked in #56025. Rounding or CSS-adjusted dimensions preserve the intrinsic aspect ratio, but the warning compares the rendered axes too literally. Closing as a duplicate so the warning fix in #92490 stays consolidated.
